src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -292,70 +298,69 @@ private BoundExpressionStatement SetDisposeMode(bool value)
         /// </summary>
         public override BoundNode VisitTryStatement(BoundTryStatement node)
         {
-            _previousDisposalLabels ??= new ArrayBuilder<LabelSymbol>();
-            _previousDisposalLabels.Push(_enclosingFinallyEntryOrFinallyExitOrExitLabel);
-
-            var finallyExit = F.GenerateLabel("finallyExit");
-            _previousDisposalLabels.Push(finallyExit);
-
-            if (node.FinallyBlockOpt != null)
+            bool addedDisposalLabel;
+            if (node.FinallyBlockOpt is object)
             {
                 var finallyEntry = F.GenerateLabel("finallyEntry");
-                _enclosingFinallyEntryOrFinallyExitOrExitLabel = finallyEntry;
+                _disposalLabels.Add(finallyEntry);
+                addedDisposalLabel = true;
 
                 // Add finallyEntry label:
                 //  try
                 //  {
                 //      ...
                 //      finallyEntry:
-
-                // Add finallyExit label:
-                //  finally
-                //  {
-                //      ...
-                //      finallyExit:
                 //  }
+
                 node = node.Update(
-                    tryBlock: F.Block(node.TryBlock, F.Label(finallyEntry)),
-                    node.CatchBlocks, F.Block(node.FinallyBlockOpt, F.Label(finallyExit)), node.FinallyLabelOpt, node.PreferFaultHandler);
+                     tryBlock: F.Block(node.TryBlock, F.Label(finallyEntry)),
+                    node.CatchBlocks, node.FinallyBlockOpt, node.FinallyLabelOpt, node.PreferFaultHandler);
+            }
+            else if (node.FinallyLabelOpt is object)
+            {
+                _disposalLabels.Add(node.FinallyLabelOpt);
+                addedDisposalLabel = true;
             }
-            else if ((object)node.FinallyLabelOpt != null)
+            else
             {
-                _enclosingFinallyEntryOrFinallyExitOrExitLabel = node.FinallyLabelOpt;
+                addedDisposalLabel = false;
             }
 
             var result = (BoundStatement)base.VisitTryStatement(node);
 
-            // While inside the try and catch blocks, we'll use the current finallyEntry label for disposal
-            // As soon as we close the try and catch blocks (ie. possibly enter the finally block), we'll use the finallyExit label for disposal (restored/popped from the stack by CloseTryCatchBlocks)
-            Debug.Assert(_enclosingFinallyEntryOrFinallyExitOrExitLabel == finallyExit);
-
-            // When exiting the try statement, we restore the previous disposal label.
-            _enclosingFinallyEntryOrFinallyExitOrExitLabel = _previousDisposalLabels.Pop();
+            if (addedDisposalLabel)
+            {
+                _disposalLabels.Pop();
+            }
 
-            if (node.FinallyBlockOpt != null)
+            if (node.FinallyBlockOpt != null && CurrentDisposalLabel is object)
```

Comment:
:memo: this additional condition is the core of the change. 
We don't want to generate a jump when inside a `finally`. We achieve that by stacking a `null` disposal label when visiting `finally` blocks (both real and extracted ones).

Reply:
That last line "We don't want... We achieve..." would be a great comment here.

---
In reply to: [458340690](https://github.com/dotnet/roslyn/pull/46188#discussion_r458340690) [](ancestors = 458340690)